File: javascript/poker/poker.js
const [HIGH_CARD, ONE_PAIR, TWO_PAIR, THREE_OF_A_KIND, STRAIGHT, FLUSH, FULL_HOUSE, FOUR_OF_A_KIND, STRAIGHT_FLUSH] = [0, 1, 2, 3, 4, 5, 6, 7, 8];

export const bestHands = (hands) => {
    if (hands.length === 1) {
        return hands;
    }

    let highestClassification = hands.map(classifyHand).toSorted((a, b) => b - a)[0];

    let highestHands = hands.filter(hand => classifyHand(hand) === highestClassification);

    if (highestHands.length === 1) {
        return highestHands;
    }

    return breakTie(highestHands, highestClassification);

};

const classifyHand = (hand) => {

    let parsedHand = parseHand(hand);

    if (isStraight(parsedHand) && isFlush(parsedHand)) {
        return STRAIGHT_FLUSH;
    }

    let freqs = frequencies(parsedHand.map(c => c.rank)).toString();

    if (freqs === '4,1') {
        return FOUR_OF_A_KIND;
    }

    if (freqs === '3,2') {
        return FULL_HOUSE;
    }

    if (isFlush(parsedHand)) {
        return FLUSH;
    }

    if (isStraight(parsedHand)) {
        return STRAIGHT;
    }

    if (freqs === '3,1,1') {
        return THREE_OF_A_KIND;
    }

    if (freqs === '2,2,1') {
        return TWO_PAIR;
    }

    if (freqs === '2,1,1,1') {
        return ONE_PAIR;
    }

    return HIGH_CARD;
};

const breakTie = (hands, classification) => {

    if (classification === FOUR_OF_A_KIND || classification === FULL_HOUSE) {
        return breakTieFH4AK(hands);
    }

    if (classification === STRAIGHT) {
        return breakTieStraight(hands);
    }

    return sortHighCard(hands, 0);
};

const sortHighCard = (hands, pos) => {

    if (pos === 5 || hands.length === 1) {
        return hands;
    }

    let highestNthCards = hands.map(hand => sortCards(parseHand(hand))[pos]);
    let highestNthRank = sortCards(highestNthCards)[0].rank;

    let filteredHands = hands.filter(hand => sortCards(parseHand(hand))[pos].rank === highestNthRank);

    return sortHighCard(filteredHands, pos + 1);

};

const parseHand = (hand) => hand.split(' ').map(parseCard);

const parseCard = (card) => {
    let rank = 0;
    let {r, s} = /^(?<r>\d+|[AJQK])(?<s>[SHDC])$/.exec(card).groups;

    switch(r) {
        case 'A':
            rank = 14;
            break;
        case 'J':
            rank = 11;
            break;
        case 'Q':
            rank = 12;
            break;
        case 'K':
            rank = 13;
            break;
        default:
            rank = Number(r);
            break;
    }

    return {rank, suit: s};
};

const sortCards = (cards) => cards.toSorted((cardA, cardB) => cardB.rank - cardA.rank);

const isFlush = (hand) => new Set(hand.map(c => c.suit)).size === 1;

const isStraight = (hand) => {
    let sortedRanks = sortCards(hand).map(c => c.rank);

    return (sortedRanks[0] === 14 && isDecreasingSequence(sortedRanks.slice(1)))
            || isDecreasingSequence(sortedRanks);
};

const isDecreasingSequence = (numbers) => 
    numbers.length < 2 
        || ((numbers[0] - numbers[1] === 1) && isDecreasingSequence(numbers.slice(1)));

const frequencies = (numbers) => 
        [...numbers
            .reduce((acc, number) => acc.set(number, (acc.get(number) || 0) + 1), new Map())
            .values()].toSorted((a, b) => b - a);

const breakTieFH4AK = (hands) => {
    const sumOfHand = (hand) => parseHand(hand).reduce((acc, card) => acc + card.rank, 0);
    return [hands.toSorted((handA, handB) => sumOfHand(handB) - sumOfHand(handA))[0]];
};

const breakTieStraight = (hands) => {
    const secondCard = (hand) => sortCards(parseHand(hand))[1].rank;
    return [hands.toSorted((handA, handB) => secondCard(handB) - secondCard(handA))[0]];
};

File: elixir/poker/lib/poker.ex
defmodule Poker do
  @doc """
  Given a list of poker hands, return a list containing the highest scoring hand.

  If two or more hands tie, return the list of tied hands in the order they were received.

  The basic rules and hand rankings for Poker can be found at:

    https://en.wikipedia.org/wiki/List_of_poker_hands

  For this exercise, we'll consider the game to be using no Jokers,
  so five-of-a-kind hands will not be tested. We will also consider
  the game to be using multiple decks, so it is possible for multiple
  players to have identical cards.

  Aces can be used in low (A 2 3 4 5) or high (10 J Q K A) straights, but do not count as
  a high card in the former case.

  For example, (A 2 3 4 5) will lose to (2 3 4 5 6).

  You can also assume all inputs will be valid, and do not need to perform error checking
  when parsing card values. All hands will be a list of 5 strings, containing a number
  (or letter) for the rank, followed by the suit.

  Ranks (lowest to highest): 2 3 4 5 6 7 8 9 10 J Q K A
  Suits (order doesn't matter): C D H S

  Example hand: ~w(4S 5H 4C 5D 4H) # Full house, 5s over 4s
  """

  @spec best_hand(list(list(String.t()))) :: list(list(String.t()))
  def best_hand(hands) when length(hands) == 1, do: hands

  def best_hand(hands) do
    hands
    |> Enum.map(&Hand.parse_hand/1)
    |> find_best_hand()
  end

  defp find_best_hand(hands) do
    sorted_hands =
      hands
      |> Hand.sort_hands()

    winning_category = sorted_hands |> hd() |> Map.get(:category)

    winning_hands = sorted_hands |> Enum.filter(fn hand -> hand.category == winning_category end)

    case winning_hands |> length() do
      1 -> winning_hands |> Enum.map(&Hand.print_hand/1)
      _ -> find_high_hands(winning_hands, winning_category) |> Enum.map(&Hand.print_hand/1)
    end
  end

  defp find_high_hands(hands, category) do
    case category do
      :straight_flush -> tie_breaker_hc(hands)
      :four_of_a_kind -> tie_breaker_foak(hands)
      :full_house -> tie_breaker_full_house(hands)
      :flush -> tie_breaker_hc(hands)
      :five_high_straight -> hands
      :ace_high_straight -> hands
      :straight -> tie_breaker_straight(hands)
      :three_of_a_kind -> tie_breaker_toak(hands)
      :two_pair -> tie_breaker_op(hands)
      :one_pair -> tie_breaker_op(hands)
      :high_card -> tie_breaker_hc(hands)
    end
  end

  defp tie_breaker_foak([first_hand | hands]), do: tie_breaker_foak(hands, [first_hand])

  defp tie_breaker_foak([], best_hands) do
    case best_hands |> length() do
      1 -> best_hands
      _ -> tie_breaker_hc(best_hands)
    end
  end

  defp tie_breaker_foak([hand | hands], [best_hand | _tail] = best_hands) do
    quad_0 = get_group(hand, 4) |> Card.sort_cards()
    quad_1 = get_group(best_hand, 4) |> Card.sort_cards()

    case get_high_card(quad_0, quad_1) do
      :tie -> tie_breaker_foak(hands, [hand | best_hands])
      :first -> tie_breaker_foak(hands, [hand])
      :second -> tie_breaker_foak(hands, best_hands)
    end
  end

  defp tie_breaker_full_house([first_hand | hands]),
    do: tie_breaker_full_house(hands, [first_hand])

  defp tie_breaker_full_house([], best_hands) do
    case best_hands |> length() do
      1 -> best_hands
      _ -> tie_breaker_op(best_hands)
    end
  end

  defp tie_breaker_full_house([hand | hands], [best_hand | _tail] = best_hands) do
    triplet_0 = get_group(hand, 3) |> Card.sort_cards()
    triplet_1 = get_group(best_hand, 3) |> Card.sort_cards()

    case get_high_card(triplet_0, triplet_1) do
      :tie -> tie_breaker_full_house(hands, [hand | best_hands])
      :first -> tie_breaker_full_house(hands, [hand])
      :second -> tie_breaker_full_house(hands, best_hands)
    end
  end

  defp tie_breaker_straight([first_hand | hands]), do: tie_breaker_straight(hands, [first_hand])

  defp tie_breaker_straight([], best_hands), do: best_hands

  defp tie_breaker_straight([hand | hands], [best_hand | _tail] = best_hands) do
    h_0 = hand.cards |> Card.sort_cards()
    h_1 = best_hand.cards |> Card.sort_cards()

    case get_high_card(h_0, h_1) do
      :tie -> tie_breaker_straight(hands, [hand | best_hands])
      :first -> tie_breaker_straight(hands, [hand])
      :second -> tie_breaker_straight(hands, best_hands)
    end
  end

  defp tie_breaker_toak([first_hand | hands]), do: tie_breaker_toak(hands, [first_hand])

  defp tie_breaker_toak([], best_hands) do
    case best_hands |> length() do
      1 -> best_hands
      _ -> tie_breaker_hc(best_hands)
    end
  end

  defp tie_breaker_toak([hand | hands], [best_hand | _tail] = best_hands) do
    triplet_0 = get_group(hand, 3) |> Card.sort_cards()
    triplet_1 = get_group(best_hand, 3) |> Card.sort_cards()

    case get_high_card(triplet_0, triplet_1) do
      :tie -> tie_breaker_toak(hands, [hand | best_hands])
      :first -> tie_breaker_toak(hands, [hand])
      :second -> tie_breaker_toak(hands, best_hands)
    end
  end

  defp tie_breaker_op([first_hand | hands]), do: tie_breaker_op(hands, [first_hand])

  defp tie_breaker_op([], best_hands) do
    case best_hands |> length() do
      1 -> best_hands
      _ -> tie_breaker_hc(best_hands)
    end
  end

  defp tie_breaker_op([hand | hands], [best_hand | _tail] = best_hands) do
    pairs_0 = get_group(hand, 2) |> Card.sort_cards()
    pairs_1 = get_group(best_hand, 2) |> Card.sort_cards()

    case get_high_card(pairs_0, pairs_1) do
      :tie -> tie_breaker_op(hands, [hand | best_hands])
      :first -> tie_breaker_op(hands, [hand])
      :second -> tie_breaker_op(hands, best_hands)
    end
  end

  defp tie_breaker_hc([first_hand | hands]), do: tie_breaker_hc(hands, [first_hand])
  defp tie_breaker_hc([], best_hands), do: best_hands

  defp tie_breaker_hc([hand | hands], [best_hand | _tail] = best_hands) do
    case get_high_card(Card.sort_cards(hand.cards), Card.sort_cards(best_hand.cards)) do
      :tie -> tie_breaker_hc(hands, [hand | best_hands])
      :first -> tie_breaker_hc(hands, [hand])
      :second -> tie_breaker_hc(hands, best_hands)
    end
  end

  defp get_high_card([], []), do: :tie

  defp get_high_card([card_0 | hand_0], [card_1 | hand_1]) when card_0.value == card_1.value,
    do: get_high_card(hand_0, hand_1)

  defp get_high_card([card_0 | _hand_0], [card_1 | _hand_1]) when card_0.value > card_1.value,
    do: :first

  defp get_high_card([card_0 | _hand_0], [card_1 | _hand_1]) when card_0.value < card_1.value,
    do: :second

  defp get_group(hand, frequency) do
    groups =
      hand.cards
      |> Enum.map(fn card -> card.rank end)
      |> Enum.frequencies()
      |> Map.to_list()
      |> Enum.filter(fn {_rank, freq} -> freq == frequency end)
      |> Enum.map(fn {rank, _freq} -> rank end)

    hand.cards
    |> Enum.filter(fn card -> card.rank in groups end)
  end
end
